cmd/gateway-gcs.go

```diff
@@ -294,8 +282,48 @@ func (l *gcsGateway) ListBuckets() ([]BucketInfo, error) {
 	return b, nil
 }
 
-// DeleteBucket delete a bucket on GCS
+// DeleteBucket delete a bucket on GCS.
 func (l *gcsGateway) DeleteBucket(bucket string) error {
+	itObject := l.client.Bucket(bucket).Objects(l.ctx, &storage.Query{Delimiter: slashSeparator, Versions: false})
+	// We list the bucket and if we find any objects we return BucketNotEmpty error. If we
+	// find only "minio.sys.temp/" then we remove it before deleting the bucket.
+	gcsMinioPathFound := false
+	nonGCSMinioPathFound := false
+	for {
```

Comment:
> Unable to imagine how functional test will solve without seeing it.

For functional test we can just do this:
* upload an object `testobj`
* delete the bucket - should fail
* create a new multipart, upload a part
* delete the bucket - should fail
* delete the object `testobj`
* delete the bucket - should succeed

> Overall, I understand GCS gives in one format and minio converts into S3 format. Why can't we have functions which accept GCS format data types and retuning S3 data types 

We already do that if it makes sense for ex fromGCSAttrsToObjectInfo()

> and having them unit tested?

again we should not write code just so that it can be unit tested. We should write code for simplicity and readability. If we have a stateless function then we should unit test it.

Can you send a PR to this PR showing what you have in mind?